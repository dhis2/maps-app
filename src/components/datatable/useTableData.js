import i18n from '@dhis2/d2-i18n'
import { useMemo } from 'react'
import { useSelector } from 'react-redux'
import {
    EVENT_LAYER,
    THEMATIC_LAYER,
    ORG_UNIT_LAYER,
    EARTH_ENGINE_LAYER,
    FACILITY_LAYER,
    GEOJSON_URL_LAYER,
} from '../../constants/layers.js'
import { numberValueTypes } from '../../constants/valueTypes.js'
import { hasClasses, getPrecision } from '../../util/earthEngine.js'
import { filterData } from '../../util/filter.js'
import { isValidUid } from '../../util/helpers.js'
import { numberPrecision } from '../../util/numbers.js'

const ASCENDING = 'asc'

const TYPE_NUMBER = 'number'
const TYPE_STRING = 'string'
const TYPE_DATE = 'date'

const INDEX = 'index'
const NAME = 'name'
const ID = 'id'
const VALUE = 'value'
const LEGEND = 'legend'
const RANGE = 'range'
const LEVEL = 'level'
const PARENT_NAME = 'parentName'
const TYPE = 'type'
const COLOR = 'color'
const OUNAME = 'ouname'
const EVENTDATE = 'eventdate'

const defaultFieldsMap = () => ({
    [INDEX]: { name: i18n.t('Index'), dataKey: INDEX, type: TYPE_NUMBER },
    [NAME]: { name: i18n.t('Name'), dataKey: NAME, type: TYPE_STRING },
    [ID]: { name: i18n.t('Id'), dataKey: ID, type: TYPE_STRING },
    [LEVEL]: { name: i18n.t('Level'), dataKey: LEVEL, type: TYPE_NUMBER },
    [PARENT_NAME]: {
        name: i18n.t('Parent'),
        dataKey: PARENT_NAME,
        type: TYPE_STRING,
    },
    [TYPE]: { name: i18n.t('Type'), dataKey: TYPE, type: TYPE_STRING },
    [VALUE]: { name: i18n.t('Value'), dataKey: VALUE, type: TYPE_NUMBER },
    [LEGEND]: { name: i18n.t('Legend'), dataKey: LEGEND, type: TYPE_STRING },
    [RANGE]: { name: i18n.t('Range'), dataKey: RANGE, type: TYPE_STRING },
    [OUNAME]: { name: i18n.t('Org unit'), dataKey: OUNAME, type: TYPE_STRING },
    [EVENTDATE]: {
        name: i18n.t('Event time'),
        dataKey: EVENTDATE,
        type: TYPE_DATE,
        renderer: 'formatTime...',
    },
    [COLOR]: {
        name: i18n.t('Color'),
        dataKey: COLOR,
        type: TYPE_STRING,
        renderer: 'rendercolor',
    },
})

const getThematicHeaders = () =>
    [
        INDEX,
        NAME,
        ID,
        VALUE,
        LEGEND,
        RANGE,
        LEVEL,
        PARENT_NAME,
        TYPE,
        COLOR,
    ].map((field) => defaultFieldsMap()[field])

const getEventHeaders = ({ layerHeaders = [], styleDataItem }) => {
    const fields = [INDEX, OUNAME, ID, EVENTDATE].map(
        (field) => defaultFieldsMap()[field]
    )

    const customFields = layerHeaders
        .filter(({ name }) => isValidUid(name))
        .map(({ name: dataKey, column: name, valueType }) => ({
            name,
            dataKey,
            type: numberValueTypes.includes(valueType)
                ? TYPE_NUMBER
                : TYPE_STRING,
        }))

    customFields.push(defaultFieldsMap()[TYPE])

    if (styleDataItem) {
        customFields.push(defaultFieldsMap()[COLOR])
    }

    return fields.concat(customFields)
}

const getOrgUnitHeaders = () =>
    [INDEX, NAME, ID, LEVEL, PARENT_NAME, TYPE].map(
        (field) => defaultFieldsMap()[field]
    )

const getFacilityHeaders = () =>
    [INDEX, NAME, ID, TYPE].map((field) => defaultFieldsMap()[field])

const toTitleCase = (str) =>
    str.replace(
        /\w\S*/g,
        (txt) => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
    )

const getEarthEngineHeaders = ({ aggregationType, legend, data }) => {
    const { title, items } = legend

    let customFields = []

    if (hasClasses(aggregationType) && items) {
        customFields = items.map(({ id, name }) => ({
            name,
            dataKey: String(id),
            roundFn: numberPrecision(2),
            type: TYPE_NUMBER,
        }))
    } else if (Array.isArray(aggregationType) && aggregationType.length) {
        customFields = aggregationType.map((type) => {
            let roundFn = null
            if (data?.length) {
                const precision = getPrecision(data.map((d) => d[type]))
                roundFn = numberPrecision(precision)
            }
            return {
                name: toTitleCase(`${type} ${title}`),
                dataKey: type,
                roundFn,
                type: TYPE_NUMBER,
            }
        })
    }

    return [INDEX, NAME, ID, TYPE]
        .map((field) => defaultFieldsMap()[field])
        .concat(customFields)
}

const getGeoJsonUrlHeaders = (data) => {
    const customFields = Object.entries(data[0])
        .filter(
            ([, value]) =>
                typeof value === TYPE_NUMBER || typeof value === TYPE_STRING
        )
        .map(([key, value]) => {
            let roundFn = null
            const type =
                typeof value === TYPE_NUMBER ? TYPE_NUMBER : TYPE_STRING
            if (type === TYPE_NUMBER) {
                const precision = getPrecision(data.map((d) => d[key]))
                roundFn = numberPrecision(precision)
            }

            return {
                name: key,
                dataKey: key,
                type,
                roundFn,
            }
        })

    customFields.push(defaultFieldsMap()[TYPE])
    return customFields
}

const EMPTY_AGGREGATIONS = {}
const EMPTY_LAYER = {}

export const useTableData = ({ layer, sortField, sortDirection }) => {
    const allAggregations = useSelector((state) => state.aggregations)
    const aggregations = allAggregations[layer.id] || EMPTY_AGGREGATIONS

    const {
        layer: layerType,
        aggregationType,
        legend,
        styleDataItem,
        data,
        dataFilters,
        headers: layerHeaders,
        serverCluster,
    } = layer || EMPTY_LAYER

    const dataWithAggregations = useMemo(() => {
        if (!data || serverCluster) {
            return null
        }
        return data
            .map((d, i) => ({
                index: i,
                ...d,
            }))
            .filter((d) => !d.properties.hasAdditionalGeometry)
            .map((d, i) => ({
                ...(d.properties || d),
                ...aggregations[d.id],
                index: d.index,
                i,
            }))
    }, [data, aggregations, serverCluster])

    const headers = useMemo(() => {
        if (dataWithAggregations === null) {
            return null
        }

        switch (layerType) {
            case THEMATIC_LAYER:
                return getThematicHeaders()
            case EVENT_LAYER:
                return getEventHeaders({ layerHeaders, styleDataItem })
            case ORG_UNIT_LAYER:
                return getOrgUnitHeaders()
            case EARTH_ENGINE_LAYER:
                return getEarthEngineHeaders({
                    aggregationType,
                    legend,
                    data: dataWithAggregations,
                })
            case FACILITY_LAYER:
                return getFacilityHeaders()
            case GEOJSON_URL_LAYER:
                return getGeoJsonUrlHeaders(dataWithAggregations)
            default: {
                return null
            }
        }
    }, [
        layerType,
        aggregationType,
        legend,
        styleDataItem,
        dataWithAggregations,
        layerHeaders,
    ])

    const rows = useMemo(() => {
        if (dataWithAggregations === null || headers === null) {
            return null
        }

        if (!dataWithAggregations.length || !headers?.length) {
            return []
        }

        const filteredData = filterData(dataWithAggregations, dataFilters)

        //sort
        filteredData.sort((a, b) => {
            a = a[sortField]
            b = b[sortField]

            if (typeof a === TYPE_NUMBER) {
                return sortDirection === ASCENDING ? a - b : b - a
            }
            // TODO: Make sure sorting works across different locales - use lib method
            if (a !== undefined) {
                return sortDirection === ASCENDING
                    ? a.localeCompare(b)
                    : b.localeCompare(a)
            }

            return 0
        })

        return filteredData.map((item) =>
            headers.map(({ dataKey, roundFn, type }) => {
                const value = roundFn ? roundFn(item[dataKey]) : item[dataKey]

                return {
                    dataKey,
                    value: type === TYPE_NUMBER && isNaN(value) ? null : value,
                    align: type === TYPE_NUMBER ? 'right' : 'left',
                }
            })
        )
    }, [headers, dataWithAggregations, dataFilters, sortField, sortDirection])

    // EE layers and event layers may be loading additional data
    const isLoading =
        (layerType === EARTH_ENGINE_LAYER &&
            aggregationType?.length &&
            (!aggregations || aggregations === EMPTY_AGGREGATIONS)) ||
        (layerType === EVENT_LAYER && !layer.isExtended && !serverCluster)

    let error = null
    if (serverCluster) {
        error = i18n.t(
            'Data table is not supported when events are grouped on the server.'
        )
    } else if (dataWithAggregations === null) {
        error = i18n.t(
            'No valid data was found for the current layer configuration.'
        )
    } else if (headers === null) {
        error = i18n.t('Data table is not supported for this layer type.')
    }

    return { headers, rows, isLoading, error }
}
