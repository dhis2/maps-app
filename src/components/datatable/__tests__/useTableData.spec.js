import { render } from '@testing-library/react'
import PropTypes from 'prop-types'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import { useTableData } from '../useTableData.js'

jest.mock('../../map/MapApi.js', () => ({
    loadEarthEngineWorker: jest.fn(),
}))

const Table = ({ layer, sortField, sortDirection }) => {
    const { headers, rows } = useTableData({ layer, sortField, sortDirection })

    return (
        <>
            <ul className="headers">
                {headers.map((header) => (
                    <li
                        key={header.name}
                    >{`name=${header.name}, dataKey=${header.dataKey}, type=${header.type} and roundFn=${header.roundFn}`}</li>
                ))}
            </ul>
            <ul className="rows">
                {rows.map((row, index) => (
                    <ul key={index} className="row">
                        {row.map((r) => (
                            <li
                                key={r.dataKey}
                            >{`value=${r.value} and dataKey=${r.dataKey}`}</li>
                        ))}
                    </ul>
                ))}
            </ul>
        </>
    )
}

Table.propTypes = {
    layer: PropTypes.object,
    sortDirection: PropTypes.string,
    sortField: PropTypes.string,
}

const mockStore = configureMockStore()

describe('useTableData', () => {
    test('gets headers and rows for facility layer', () => {
        const store = {
            aggregations: {},
        }
        const layer = {
            layer: 'facility',
            dataFilters: null,
            data: [
                {
                    properties: {
                        id: 'facility-1',
                        name: 'Facility 1',
                        type: 'Point',
                    },
                },
            ],
        }
        const { container } = render(
            <Provider store={mockStore(store)}>
                <Table layer={layer} sortField="name" sortDirection="asc" />
            </Provider>
        )

        expect(container).toMatchSnapshot()
    })

    test('gets headers and rows for orgUnit layer', () => {
        const store = {
            aggregations: {},
        }
        const layer = {
            layer: 'orgUnit',
            dataFilters: null,
            data: [
                {
                    properties: {
                        id: 'orgunit-id-1',
                        name: 'OrgUnitName 1',
                        type: 'MultiPolygon',
                        level: 3,
                        parentName: 'Bo',
                    },
                },
            ],
        }
        const { container } = render(
            <Provider store={mockStore(store)}>
                <Table layer={layer} sortField="name" sortDirection="asc" />
            </Provider>
        )

        expect(container).toMatchSnapshot()
    })

    test('gets headers and rows for thematic layer', () => {
        const store = {
            aggregations: {},
        }
        const layer = {
            layer: 'thematic',
            dataFilters: null,
            data: [
                {
                    properties: {
                        type: 'Point',
                        id: 'thematicId-1',
                        name: 'Ngelehun CHC',
                        level: 4,
                        parentName: 'Badjia',
                        color: '#FFFFB2',
                        legend: 'Great',
                        range: '90 - 120',
                        value: 106.3,
                    },
                },
            ],
        }
        const { container } = render(
            <Provider store={mockStore(store)}>
                <Table layer={layer} sortField="name" sortDirection="asc" />
            </Provider>
        )

        expect(container).toMatchSnapshot()
    })

    test('gets headers and rows for event layer', () => {
        const store = {
            aggregations: {},
        }
        const layer = {
            layer: 'event',
            dataFilters: null,
            headers: [
                {
                    name: 'ps',
                    column: 'Program stage',
                    valueType: 'TEXT',
                },
                {
                    name: 'eventdate',
                    column: 'Report date',
                    valueType: 'DATE',
                },
                {
                    name: 'lastupdated',
                    column: 'Last updated on',
                    valueType: 'DATE',
                },
                {
                    name: 'eventstatus',
                    column: 'Event status',
                    valueType: 'TEXT',
                },
                {
                    name: 'oZg33kd9taw',
                    column: 'Gender',
                    valueType: 'TEXT',
                },
            ],

            data: [
                {
                    properties: {
                        id: 'a9712323629',
                        type: 'Point',
                        ps: 'pTo4uMt3xur',
                        eventdate: '2023-05-15 00:00:00.0',
                        lastupdated: '2018-04-12 20:58:51.31',
                        ouname: 'Lumley Hospital',
                        eventstatus: 'ACTIVE',
                        oZg33kd9taw: 'Female',
                        value: 'Female',
                        color: '#ff7f00',
                    },
                },
            ],
        }
        const { container } = render(
            <Provider store={mockStore(store)}>
                <Table layer={layer} sortField="name" sortDirection="asc" />
            </Provider>
        )

        expect(container).toMatchSnapshot()
    })

    test('gets headers and rows for EE population layer', () => {
        const store = {
            aggregations: {
                eelayerid: {
                    boOU: {
                        mean: 47.34593724212383,
                        sum: 851090.567864418,
                    },
                    bombaliOU: {
                        mean: 27.347820392739166,
                        sum: 585872.3562736511,
                    },
                },
            },
        }

        const layer = {
            layer: 'earthEngine',
            aggregationType: ['sum', 'mean'],
            legend: {
                title: 'Population',
            },
            id: 'eelayerid',
            dataFilters: null,
            data: [
                {
                    id: 'boOU',
                    properties: {
                        type: 'Polygon',
                        id: 'boOu',
                        name: 'Bo',
                    },
                },
                {
                    id: 'bombaliOU',
                    properties: {
                        type: 'Polygon',
                        id: 'bombaliOU',
                        name: 'Bombali',
                    },
                },
            ],
        }
        const { container } = render(
            <Provider store={mockStore(store)}>
                <Table layer={layer} sortField="name" sortDirection="asc" />
            </Provider>
        )

        expect(container).toMatchSnapshot()
    })

    test('gets headers and rows for EE population age groups layer', () => {
        const store = {
            aggregations: {
                eelayerid: {
                    badijaOU: {
                        M_0_mean: 0.4416957503717281,
                        M_0_sum: 279.5934099853039,
                        M_1_mean: 1.667343524395007,
                        M_1_sum: 1055.4284509420395,
                        M_5_mean: 1.8668244672235907,
                        M_5_sum: 1181.699887752533,
                        mean: 3.975863741990326,
                        sum: 2516.7217486798763,
                    },
                    baomaOU: {
                        M_0_mean: 0.6669754306043404,
                        M_0_sum: 1012.4687036573887,
                        M_1_mean: 2.517744771694477,
                        M_1_sum: 3821.9365634322166,
                        M_5_mean: 2.818359887764859,
                        M_5_sum: 4278.270309627056,
                        mean: 6.003080090063677,
                        sum: 9112.675576716661,
                    },
                },
            },
        }

        const layer = {
            layer: 'earthEngine',
            name: 'Population age groups',
            aggregationType: ['sum', 'mean'],
            id: 'eelayerid',
            legend: {
                title: 'Population age groups',
                groups: [
                    {
                        id: 'M_0',
                        name: 'Male 0 - 1 years',
                    },
                    {
                        id: 'M_1',
                        name: 'Male 1 - 4 years',
                    },
                    {
                        id: 'M_5',
                        name: 'Male 5 - 9 years',
                    },
                ],
                items: [
                    {
                        color: '#fee5d9',
                        from: 0,
                        to: 10,
                        name: '0 - 10',
                    },
                    {
                        color: '#fcbba1',
                        from: 10,
                        to: 20,
                        name: '10 - 20',
                    },
                    {
                        color: '#fc9272',
                        from: 20,
                        to: 30,
                        name: '20 - 30',
                    },
                    {
                        color: '#fb6a4a',
                        from: 30,
                        to: 40,
                        name: '30 - 40',
                    },
                    {
                        color: '#de2d26',
                        from: 40,
                        to: 50,
                        name: '40 - 50',
                    },
                    {
                        color: '#a50f15',
                        from: 50,
                        name: '> 50',
                    },
                ],
            },
            data: [
                {
                    id: 'badijaOU',
                    properties: {
                        type: 'Polygon',
                        id: 'boOU',
                        name: 'Badija',
                    },
                },
                {
                    type: 'Feature',
                    id: 'baomaOU',
                    properties: {
                        type: 'Polygon',
                        id: 'baomaOU',
                        name: 'Baoma',
                    },
                },
            ],
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <Table layer={layer} sortField="name" sortDirection="asc" />
            </Provider>
        )

        expect(container).toMatchSnapshot()
    })
})
