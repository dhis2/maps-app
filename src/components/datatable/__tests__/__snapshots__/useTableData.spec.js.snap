// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTableData gets headers and rows for EE population age groups layer 1`] = `
<div>
  <ul
    class="headers"
  >
    <li>
      name=Index, dataKey=index, type=undefined and roundFn=undefined
    </li>
    <li>
      name=Name, dataKey=name, type=string and roundFn=undefined
    </li>
    <li>
      name=Id, dataKey=id, type=string and roundFn=undefined
    </li>
    <li>
      name=Type, dataKey=type, type=string and roundFn=undefined
    </li>
    <li>
      name=Sum Population Age Groups, dataKey=sum, type=number and roundFn=n =&gt; Math.round(n * m) / m
    </li>
    <li>
      name=Mean Population Age Groups, dataKey=mean, type=number and roundFn=n =&gt; Math.round(n * m) / m
    </li>
  </ul>
  <ul
    class="rows"
  >
    <ul
      class="row"
    >
      <li>
        value=0 and dataKey=index
      </li>
      <li>
        value=Badija and dataKey=name
      </li>
      <li>
        value=boOU and dataKey=id
      </li>
      <li>
        value=Polygon and dataKey=type
      </li>
      <li>
        value=2517 and dataKey=sum
      </li>
      <li>
        value=3.976 and dataKey=mean
      </li>
    </ul>
    <ul
      class="row"
    >
      <li>
        value=1 and dataKey=index
      </li>
      <li>
        value=Baoma and dataKey=name
      </li>
      <li>
        value=baomaOU and dataKey=id
      </li>
      <li>
        value=Polygon and dataKey=type
      </li>
      <li>
        value=9113 and dataKey=sum
      </li>
      <li>
        value=6.003 and dataKey=mean
      </li>
    </ul>
  </ul>
</div>
`;

exports[`useTableData gets headers and rows for EE population layer 1`] = `
<div>
  <ul
    class="headers"
  >
    <li>
      name=Index, dataKey=index, type=undefined and roundFn=undefined
    </li>
    <li>
      name=Name, dataKey=name, type=string and roundFn=undefined
    </li>
    <li>
      name=Id, dataKey=id, type=string and roundFn=undefined
    </li>
    <li>
      name=Type, dataKey=type, type=string and roundFn=undefined
    </li>
    <li>
      name=Sum Population, dataKey=sum, type=number and roundFn=n =&gt; Math.round(n * m) / m
    </li>
    <li>
      name=Mean Population, dataKey=mean, type=number and roundFn=n =&gt; Math.round(n * m) / m
    </li>
  </ul>
  <ul
    class="rows"
  >
    <ul
      class="row"
    >
      <li>
        value=0 and dataKey=index
      </li>
      <li>
        value=Bo and dataKey=name
      </li>
      <li>
        value=boOu and dataKey=id
      </li>
      <li>
        value=Polygon and dataKey=type
      </li>
      <li>
        value=851091 and dataKey=sum
      </li>
      <li>
        value=47.35 and dataKey=mean
      </li>
    </ul>
    <ul
      class="row"
    >
      <li>
        value=1 and dataKey=index
      </li>
      <li>
        value=Bombali and dataKey=name
      </li>
      <li>
        value=bombaliOU and dataKey=id
      </li>
      <li>
        value=Polygon and dataKey=type
      </li>
      <li>
        value=585872 and dataKey=sum
      </li>
      <li>
        value=27.35 and dataKey=mean
      </li>
    </ul>
  </ul>
</div>
`;

exports[`useTableData gets headers and rows for event layer 1`] = `
<div>
  <ul
    class="headers"
  >
    <li>
      name=Index, dataKey=index, type=undefined and roundFn=undefined
    </li>
    <li>
      name=Org unit, dataKey=ouname, type=string and roundFn=undefined
    </li>
    <li>
      name=Id, dataKey=id, type=string and roundFn=undefined
    </li>
    <li>
      name=Event time, dataKey=eventdate, type=date and roundFn=undefined
    </li>
    <li>
      name=Last updated on, dataKey=lastupdated, type=string and roundFn=undefined
    </li>
    <li>
      name=Event status, dataKey=eventstatus, type=string and roundFn=undefined
    </li>
    <li>
      name=Gender, dataKey=oZg33kd9taw, type=string and roundFn=undefined
    </li>
    <li>
      name=Type, dataKey=type, type=string and roundFn=undefined
    </li>
  </ul>
  <ul
    class="rows"
  >
    <ul
      class="row"
    >
      <li>
        value=0 and dataKey=index
      </li>
      <li>
        value=Lumley Hospital and dataKey=ouname
      </li>
      <li>
        value=a9712323629 and dataKey=id
      </li>
      <li>
        value=2023-05-15 00:00:00.0 and dataKey=eventdate
      </li>
      <li>
        value=2018-04-12 20:58:51.31 and dataKey=lastupdated
      </li>
      <li>
        value=ACTIVE and dataKey=eventstatus
      </li>
      <li>
        value=Female and dataKey=oZg33kd9taw
      </li>
      <li>
        value=Point and dataKey=type
      </li>
    </ul>
  </ul>
</div>
`;

exports[`useTableData gets headers and rows for facility layer 1`] = `
<div>
  <ul
    class="headers"
  >
    <li>
      name=Index, dataKey=index, type=undefined and roundFn=undefined
    </li>
    <li>
      name=Name, dataKey=name, type=string and roundFn=undefined
    </li>
    <li>
      name=Id, dataKey=id, type=string and roundFn=undefined
    </li>
    <li>
      name=Type, dataKey=type, type=string and roundFn=undefined
    </li>
  </ul>
  <ul
    class="rows"
  >
    <ul
      class="row"
    >
      <li>
        value=0 and dataKey=index
      </li>
      <li>
        value=Facility 1 and dataKey=name
      </li>
      <li>
        value=facility-1 and dataKey=id
      </li>
      <li>
        value=Point and dataKey=type
      </li>
    </ul>
  </ul>
</div>
`;

exports[`useTableData gets headers and rows for orgUnit layer 1`] = `
<div>
  <ul
    class="headers"
  >
    <li>
      name=Index, dataKey=index, type=undefined and roundFn=undefined
    </li>
    <li>
      name=Name, dataKey=name, type=string and roundFn=undefined
    </li>
    <li>
      name=Id, dataKey=id, type=string and roundFn=undefined
    </li>
    <li>
      name=Level, dataKey=level, type=number and roundFn=undefined
    </li>
    <li>
      name=Parent, dataKey=parentName, type=string and roundFn=undefined
    </li>
    <li>
      name=Type, dataKey=type, type=string and roundFn=undefined
    </li>
  </ul>
  <ul
    class="rows"
  >
    <ul
      class="row"
    >
      <li>
        value=0 and dataKey=index
      </li>
      <li>
        value=OrgUnitName 1 and dataKey=name
      </li>
      <li>
        value=orgunit-id-1 and dataKey=id
      </li>
      <li>
        value=3 and dataKey=level
      </li>
      <li>
        value=Bo and dataKey=parentName
      </li>
      <li>
        value=MultiPolygon and dataKey=type
      </li>
    </ul>
  </ul>
</div>
`;

exports[`useTableData gets headers and rows for thematic layer 1`] = `
<div>
  <ul
    class="headers"
  >
    <li>
      name=Index, dataKey=index, type=undefined and roundFn=undefined
    </li>
    <li>
      name=Name, dataKey=name, type=string and roundFn=undefined
    </li>
    <li>
      name=Id, dataKey=id, type=string and roundFn=undefined
    </li>
    <li>
      name=Value, dataKey=value, type=number and roundFn=undefined
    </li>
    <li>
      name=Legend, dataKey=legend, type=string and roundFn=undefined
    </li>
    <li>
      name=Range, dataKey=range, type=string and roundFn=undefined
    </li>
    <li>
      name=Level, dataKey=level, type=number and roundFn=undefined
    </li>
    <li>
      name=Parent, dataKey=parentName, type=string and roundFn=undefined
    </li>
    <li>
      name=Type, dataKey=type, type=string and roundFn=undefined
    </li>
    <li>
      name=Color, dataKey=color, type=string and roundFn=undefined
    </li>
  </ul>
  <ul
    class="rows"
  >
    <ul
      class="row"
    >
      <li>
        value=0 and dataKey=index
      </li>
      <li>
        value=Ngelehun CHC and dataKey=name
      </li>
      <li>
        value=thematicId-1 and dataKey=id
      </li>
      <li>
        value=106.3 and dataKey=value
      </li>
      <li>
        value=Great and dataKey=legend
      </li>
      <li>
        value=90 - 120 and dataKey=range
      </li>
      <li>
        value=4 and dataKey=level
      </li>
      <li>
        value=Badjia and dataKey=parentName
      </li>
      <li>
        value=Point and dataKey=type
      </li>
      <li>
        value=#FFFFB2 and dataKey=color
      </li>
    </ul>
  </ul>
</div>
`;
